File: crates/lvz-context/src/anchor.rs
//! Hash-anchored edits (`RECIPE.md` §6.1): address a line by a short stable hash of its
//! content instead of resending the whole file. The model reads anchored lines once, then
//! targets edits by anchor — no full-file round-trips, and an edit that no longer matches is
//! rejected rather than silently misapplied.

use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

/// Compute the anchor for a single line: 8 hex chars over the line's trimmed content.
///
/// Trimming makes the anchor insensitive to surrounding indentation churn; identical content
/// yields identical anchors (and is therefore ambiguous to target — by design).
pub fn anchor_of(line: &str) -> String {
    let mut hasher = DefaultHasher::new();
    line.trim().hash(&mut hasher);
    format!("{:08x}", hasher.finish() as u32)
}

/// A line paired with its anchor, as presented to the model.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AnchoredLine {
    pub anchor: String,
    pub text: String,
}

/// Annotate every line of `source` with its anchor.
pub fn anchored_lines(source: &str) -> Vec<AnchoredLine> {
    source
        .lines()
        .map(|text| AnchoredLine {
            anchor: anchor_of(text),
            text: text.to_string(),
        })
        .collect()
}

/// Render `source` with a leading `anchor│ ` gutter on each line — the form the model reads
/// before issuing [`Edit`]s.
pub fn render_anchored(source: &str) -> String {
    anchored_lines(source)
        .iter()
        .map(|l| format!("{}\u{2502} {}", l.anchor, l.text))
        .collect::<Vec<_>>()
        .join("\n")
}

/// What to do at a matched anchor.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum EditOp {
    /// Replace the matched line with these lines.
    Replace(String),
    /// Insert these lines immediately after the matched line.
    InsertAfter(String),
    /// Insert these lines immediately before the matched line.
    InsertBefore(String),
    /// Delete the matched line.
    Delete,
}

/// A single anchored edit.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Edit {
    pub anchor: String,
    pub op: EditOp,
}

impl Edit {
    pub fn replace(anchor: impl Into<String>, text: impl Into<String>) -> Self {
        Edit {
            anchor: anchor.into(),
            op: EditOp::Replace(text.into()),
        }
    }
    pub fn insert_after(anchor: impl Into<String>, text: impl Into<String>) -> Self {
        Edit {
            anchor: anchor.into(),
            op: EditOp::InsertAfter(text.into()),
        }
    }
    pub fn insert_before(anchor: impl Into<String>, text: impl Into<String>) -> Self {
        Edit {
            anchor: anchor.into(),
            op: EditOp::InsertBefore(text.into()),
        }
    }
    pub fn delete(anchor: impl Into<String>) -> Self {
        Edit {
            anchor: anchor.into(),
            op: EditOp::Delete,
        }
    }
}

/// Why an anchored edit could not be applied.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum AnchorError {
    /// No line matched the anchor — the file changed under the edit.
    NotFound(String),
    /// More than one line matched the anchor — the target is ambiguous.
    Ambiguous { anchor: String, count: usize },
}

impl std::fmt::Display for AnchorError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            AnchorError::NotFound(a) => write!(f, "no line matches anchor {a}"),
            AnchorError::Ambiguous { anchor, count } => {
                write!(f, "anchor {anchor} matches {count} lines (ambiguous)")
            }
        }
    }
}

impl std::error::Error for AnchorError {}

/// Apply a batch of anchored edits to `source`, preserving a trailing newline if present.
///
/// All anchors are resolved against the **original** line set, so edits don't interfere with
/// each other's targeting. Any unmatched or ambiguous anchor fails the whole batch (atomic),
/// so a stale edit never corrupts the file.
pub fn apply_edits(source: &str, edits: &[Edit]) -> Result<String, AnchorError> {
    let lines: Vec<&str> = source.lines().collect();

    // Resolve every anchor to a unique line index first.
    let mut resolved: Vec<(usize, &EditOp)> = Vec::with_capacity(edits.len());
    for edit in edits {
        let matches: Vec<usize> = lines
            .iter()
            .enumerate()
            .filter(|(_, l)| anchor_of(l) == edit.anchor)
            .map(|(i, _)| i)
            .collect();
        match matches.as_slice() {
            [] => return Err(AnchorError::NotFound(edit.anchor.clone())),
            [i] => resolved.push((*i, &edit.op)),
            many => {
                return Err(AnchorError::Ambiguous {
                    anchor: edit.anchor.clone(),
                    count: many.len(),
                })
            }
        }
    }

    // Build the output line list, consulting edits keyed by original index.
    let mut by_index: std::collections::HashMap<usize, Vec<&EditOp>> =
        std::collections::HashMap::new();
    for (i, op) in resolved {
        by_index.entry(i).or_default().push(op);
    }

    let mut out: Vec<String> = Vec::with_capacity(lines.len());
    for (i, line) in lines.iter().enumerate() {
        let ops = by_index.get(&i);
        // Inserts-before first.
        if let Some(ops) = ops {
            for op in ops {
                if let EditOp::InsertBefore(text) = op {
                    out.extend(text.lines().map(String::from));
                }
            }
        }
        // The line itself: replaced, deleted, or kept.
        let mut emitted = false;
        if let Some(ops) = ops {
            for op in ops {
                match op {
                    EditOp::Replace(text) => {
                        out.extend(text.lines().map(String::from));
                        emitted = true;
                    }
                    EditOp::Delete => emitted = true,
                    _ => {}
                }
            }
        }
        if !emitted {
            out.push((*line).to_string());
        }
        // Inserts-after last.
        if let Some(ops) = ops {
            for op in ops {
                if let EditOp::InsertAfter(text) = op {
                    out.extend(text.lines().map(String::from));
                }
            }
        }
    }

    let mut result = out.join("\n");
    if source.ends_with('\n') {
        result.push('\n');
    }
    Ok(result)
}

#[cfg(test)]
mod tests {
    use super::*;

    const SRC: &str = "fn main() {\n    let x = 1;\n    println!(\"{x}\");\n}\n";

    #[test]
    fn anchor_is_indentation_insensitive_and_stable() {
        assert_eq!(anchor_of("    let x = 1;"), anchor_of("let x = 1;"));
        assert_eq!(anchor_of("let x = 1;").len(), 8);
    }

    #[test]
    fn replace_targets_the_anchored_line() {
        let anchor = anchor_of("    let x = 1;");
        let out = apply_edits(SRC, &[Edit::replace(anchor, "    let x = 42;")]).unwrap();
        assert!(out.contains("let x = 42;"));
        assert!(!out.contains("let x = 1;"));
        assert!(out.ends_with('\n'));
    }

    #[test]
    fn insert_after_and_before() {
        let anchor = anchor_of("    let x = 1;");
        let out = apply_edits(SRC, &[Edit::insert_after(&anchor, "    let y = 2;")]).unwrap();
        let lines: Vec<&str> = out.lines().collect();
        let xi = lines.iter().position(|l| l.contains("let x = 1;")).unwrap();
        assert!(lines[xi + 1].contains("let y = 2;"));
    }

    #[test]
    fn delete_removes_the_line() {
        let anchor = anchor_of("    println!(\"{x}\");");
        let out = apply_edits(SRC, &[Edit::delete(anchor)]).unwrap();
        assert!(!out.contains("println!"));
    }

    #[test]
    fn unmatched_anchor_fails_the_batch() {
        let err = apply_edits(SRC, &[Edit::replace("deadbeef", "x")]).unwrap_err();
        assert!(matches!(err, AnchorError::NotFound(_)));
    }

    #[test]
    fn ambiguous_anchor_is_rejected() {
        let src = "dup\ndup\n";
        let err = apply_edits(src, &[Edit::replace(anchor_of("dup"), "x")]).unwrap_err();
        assert!(matches!(err, AnchorError::Ambiguous { count: 2, .. }));
    }

    #[test]
    fn render_anchored_has_gutter() {
        let rendered = render_anchored("hello");
        assert!(rendered.starts_with(&anchor_of("hello")));
        assert!(rendered.contains('\u{2502}'));
    }
}

File: src/Lavoisier/Context/Anchor.hs
{- | Hash-anchored edits (ports Rust @lvz-context::anchor@): address a line by a short stable hash of
its content instead of resending the whole file. The model reads anchored lines once, then targets
edits by anchor — no full-file round-trips, and an edit that no longer matches is rejected rather
than silently misapplied.

The anchor is an 8-hex FNV-1a hash of the line's trimmed content. It need not match the Rust
implementation's @DefaultHasher@ value — only be deterministic and self-consistent within a run —
so a dependency-free FNV-1a stands in for Rust's (unspecified, non-portable) SipHash.
-}
module Lavoisier.Context.Anchor (
    anchorOf,
    AnchoredLine (..),
    anchoredLines,
    renderAnchored,
    EditOp (..),
    Edit (..),
    replaceEdit,
    insertAfterEdit,
    insertBeforeEdit,
    deleteEdit,
    after,
    AnchorError (..),
    renderAnchorError,
    applyEdits,
)
where

import Data.Bits (shiftR, xor, (.&.))
import Data.Text (Text)
import Data.Text.Encoding (encodeUtf8)
import Data.Word (Word32)

import Data.ByteString qualified as BS
import Data.Map.Strict qualified as Map
import Data.Text qualified as T

{- | Compute the anchor for a single line: 8 hex chars over the line's trimmed content.

Trimming makes the anchor insensitive to surrounding indentation churn; identical content yields
identical anchors (and is therefore ambiguous to target — by design).
-}
anchorOf ∷ Text → Text
anchorOf = toHex8 . fnv1a . encodeUtf8 . T.strip

fnv1a ∷ BS.ByteString → Word32
fnv1a = BS.foldl' (\h b → (h `xor` fromIntegral b) * 0x01000193) 0x811c9dc5

-- | An unsigned 32-bit value as exactly 8 lowercase hex digits.
toHex8 ∷ Word32 → Text
toHex8 w = T.pack [hexDigit (fromIntegral ((w `shiftR` s) .&. 0xf)) | s ← [28, 24 .. 0]]
    where
        hexDigit d = "0123456789abcdef" !! d

-- | A line paired with its anchor, as presented to the model.
data AnchoredLine = AnchoredLine
    { alAnchor ∷ Text
    -- ^ The line's 8-hex content anchor.
    , alText ∷ Text
    -- ^ The line's raw text.
    }
    deriving stock (Eq, Show)

-- | Annotate every line of @source@ with its anchor.
anchoredLines ∷ Text → [AnchoredLine]
anchoredLines source = [AnchoredLine (anchorOf l) l | l ← T.lines source]

{- | Render @source@ with a leading @anchor│ @ gutter on each line — the form the model reads before
issuing 'Edit's.
-}
renderAnchored ∷ Text → Text
renderAnchored source =
    T.intercalate "\n" [alAnchor l <> "\9474 " <> alText l | l ← anchoredLines source]

-- | What to do at a matched anchor.
data EditOp
    = -- | Replace the matched line with these lines.
      Replace Text
    | -- | Insert these lines immediately after the matched line.
      InsertAfter Text
    | -- | Insert these lines immediately before the matched line.
      InsertBefore Text
    | -- | Delete the matched line.
      Delete
    deriving stock (Eq, Show)

{- | A single anchored edit: the anchor of the line to act on, an optional /landmark/ anchor that
disambiguates when several lines share that content, and what to do there.

'editAfter' is what makes a repeated line addressable without ever becoming positional: the target
stays content-addressed, and the landmark only says which side of a unique nearby line the meant
copy sits on — the __first__ matching line strictly after it. A line number would silently hit the
wrong line once the file shifts; this cannot. It matches @str_replace@'s @after@ argument, so one
rule covers both edit tools.
-}
data Edit = Edit
    { editAnchor ∷ Text
    , editAfter ∷ Maybe Text
    , editOp ∷ EditOp
    }
    deriving stock (Eq, Show)

-- | Replace the anchored line with @text@.
replaceEdit ∷ Text → Text → Edit
replaceEdit a t = Edit a Nothing (Replace t)

-- | Insert @text@ immediately after the anchored line.
insertAfterEdit ∷ Text → Text → Edit
insertAfterEdit a t = Edit a Nothing (InsertAfter t)

-- | Insert @text@ immediately before the anchored line.
insertBeforeEdit ∷ Text → Text → Edit
insertBeforeEdit a t = Edit a Nothing (InsertBefore t)

-- | Delete the anchored line.
deleteEdit ∷ Text → Edit
deleteEdit a = Edit a Nothing Delete

{- | Qualify an edit with the anchor of a unique landmark line: the target becomes the first matching
line after it. Disambiguates a repeated target without naming a position.
-}
after ∷ Text → Edit → Edit
after n e = e {editAfter = Just n}

{- | Why an anchored edit could not be applied. Every case is a refusal, never a guess: an edit that
cannot be pinned to exactly one line is rejected rather than applied somewhere plausible.
-}
data AnchorError
    = -- | No line matched the anchor — the file changed under the edit.
      NotFound Text
    | -- | More than one line matched the anchor (with the match count) and no @after@ was given.
      Ambiguous Text Int
    | -- | An @after@ landmark was given but matched no line.
      AfterNotFound Text
    | -- | An @after@ landmark is itself repeated (with the match count), so it pins nothing.
      AfterAmbiguous Text Int
    | -- | No line matched the anchor after the landmark (target anchor, landmark anchor).
      NoneAfter Text Text
    deriving stock (Eq, Show)

{- | A model-facing rendering of an 'AnchorError' that names the fix, not just the fault. The
@read_anchored@ gutter already shows every neighbouring anchor, so the suggested @after@ landmark
can be supplied without re-reading the file.
-}
renderAnchorError ∷ AnchorError → Text
renderAnchorError = \case
    NotFound a →
        "anchor \"" <> a <> "\" matched no line — the file changed since it was read; re-read it with read_anchored"
    Ambiguous a n →
        "anchor \""
            <> a
            <> "\" matches "
            <> tshow n
            <> " identical lines — add \"after\": the anchor of a unique line just above the one you mean"
    AfterNotFound a →
        "after-anchor \"" <> a <> "\" matched no line — re-read the file with read_anchored"
    AfterAmbiguous a n →
        "after-anchor \""
            <> a
            <> "\" is itself repeated ("
            <> tshow n
            <> " lines) — \"after\" must be unique; pick a distinctive line above the target"
    NoneAfter a n →
        "no line matching anchor \""
            <> a
            <> "\" follows after-anchor \""
            <> n
            <> "\" — \"after\" must name a line above the target"
    where
        tshow ∷ Int → Text
        tshow = T.pack . show

{- | Apply a batch of anchored edits to @source@, preserving a trailing newline if present.

All anchors are resolved against the __original__ line set, so edits don't interfere with each
other's targeting. Any unmatched or ambiguous anchor fails the whole batch (atomic), so a stale
edit never corrupts the file.
-}
applyEdits ∷ Text → [Edit] → Either AnchorError Text
applyEdits source edits = do
    let indexed = zip [0 ∷ Int ..] (T.lines source)
    -- Resolve every anchor to a unique line index first (short-circuits on the first bad anchor).
    resolved ← mapM (resolveOne indexed) edits
    let byIndex = Map.fromListWith (\new old → old <> new) [(i, [op]) | (i, op) ← resolved]
        out = concat [emit (Map.findWithDefault [] i byIndex) line | (i, line) ← indexed]
        result = T.intercalate "\n" out
    pure (if "\n" `T.isSuffixOf` source then result <> "\n" else result)
    where
        resolveOne indexed e =
            case [i | (i, l) ← indexed, anchorOf l == editAnchor e] of
                [i] → Right (i, editOp e)
                [] → Left (NotFound (editAnchor e))
                cands → case editAfter e of
                    Nothing → Left (Ambiguous (editAnchor e) (length cands))
                    Just n → case [i | (i, l) ← indexed, anchorOf l == n] of
                        [] → Left (AfterNotFound n)
                        [ni] → case [i | i ← cands, i > ni] of
                            (i : _) → Right (i, editOp e)
                            [] → Left (NoneAfter (editAnchor e) n)
                        ns → Left (AfterAmbiguous n (length ns))

        -- The output lines for one original line: inserts-before, then the line (replaced/deleted/kept),
        -- then inserts-after — mirroring the Rust ordering.
        emit ops line = pre <> middle <> post
            where
                pre = concat [T.lines t | InsertBefore t ← ops]
                post = concat [T.lines t | InsertAfter t ← ops]
                replaced = concat [T.lines t | Replace t ← ops]
                removedOrReplaced = any isReplaceOrDelete ops
                middle = if removedOrReplaced then replaced else [line]

        isReplaceOrDelete (Replace _) = True
        isReplaceOrDelete Delete = True
        isReplaceOrDelete _ = False
